File: templates/python/code-carl/ch09_dp/t28_lc0300_length_of_lis.py
"""LeetCode 300. Longest Increasing Subsequence
https://leetcode.cn/problems/longest-increasing-subsequence/
"""

from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]) == 4
    assert solution.lengthOfLIS([0, 1, 0, 3, 2, 3]) == 4
    assert solution.lengthOfLIS([7, 7, 7, 7, 7, 7, 7]) == 1
    print("All tests passed.")

File: solutions/java/code-carl/ch09_dp/T28_LC0300_LengthOfLis.java
/**
 * Practice template for:
 * - LeetCode 300. Longest Increasing Subsequence
 *   https://leetcode.cn/problems/longest-increasing-subsequence/
 */
import java.util.Arrays;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/24 23:00
 * @comment
 */
public class T28_LC0300_LengthOfLis {
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, 1);
        int maxLen = 1;
        for (int i = 1; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            maxLen = Math.max(maxLen, dp[i]);
        }
        return maxLen;
    }
}
